pkg/controller/versions/template.go

```diff
@@ -69,58 +70,89 @@ func mergeValues(base map[string]interface{}, input map[string]interface{}) map[
 	return base
 }
 
-func (v version) getSMCPTemplate(name string) (v1.ControlPlaneSpec, error) {
+func (v version) getSMCPProfile(name string) (*v1.ControlPlaneSpec, []string, error) {
 	if strings.Contains(name, "/") {
-		return v1.ControlPlaneSpec{}, fmt.Errorf("template name contains invalid character '/'")
+		return nil, nil, fmt.Errorf("profile name contains invalid character '/'")
 	}
 
-	templateContent, err := ioutil.ReadFile(path.Join(v.GetUserTemplatesDir(), name))
+	profileContent, err := ioutil.ReadFile(path.Join(v.GetUserTemplatesDir(), name))
 	if err != nil {
-		//if we can't read from the user template path, try from the default path
+		//if we can't read from the user profile path, try from the default path
 		//we use two paths because Kubernetes will not auto-update volume mounted
 		//configmaps mounted in directories with pre-existing content
-		defaultTemplateContent, defaultErr := ioutil.ReadFile(path.Join(v.GetDefaultTemplatesDir(), name))
+		defaultProfileContent, defaultErr := ioutil.ReadFile(path.Join(v.GetDefaultTemplatesDir(), name))
 		if defaultErr != nil {
-			return v1.ControlPlaneSpec{}, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
+			return nil, nil, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
 		}
-		templateContent = defaultTemplateContent
+		profileContent = defaultProfileContent
 	}
 
-	var template v1.ServiceMeshControlPlane
-	if err = yaml.Unmarshal(templateContent, &template); err != nil {
-		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+	obj, gvk, err := decoder.Decode(profileContent, nil, nil)
```

Comment:
Is there any advantage to using this over the ghodds? Ghodss is simpler, and doesn't require the additional weight of using reflection. 

Looking at https://github.com/kubernetes/kubernetes/pull/74378, it looks like Kubernetes forked ghodss, but it's unclear why.

I'd be curious if making this change slows us down/what the cost/benefit of the additional complexity is.

Reply:
It looks like they may have switched to a fork to try to support YAML 1.2, which it looks like they may be falling back on. YAML 1.2 introduces breaking changes in handling of booleans, and some other changes. Not sure how that would affect those writing templates. https://github.com/kubernetes/kubernetes/issues/34146

Reply:
No, because we don't know which version of SMCP is being loaded.  We need to pass an object to the marshaller and we don't know which one to pass (v1 or v2).

Reply:
Ah ok. Other option here would be: 

v2 := &v1.SMCP
if Unmarshal(v2) != error ... 

else unmarshal(v1) != error ...
else fail

Probably faster just to take the hit for reflection at that point though.